aptos-move/framework/aptos-framework/sources/storage_gas.move

```diff
@@ -0,0 +1,349 @@
+                } else {
+                    vector::borrow(points, i + 1)
+                }
+            )
+        };
+        curve.min_gas + (curve.max_gas - curve.min_gas) * (left.y + (current_usage_bps - left.x) * (right.y - left.y) / (right.x - left.x)) / BASIS_POINT_DENOMINATION
+    }
+
+    fun calculate_read_gas(config: &UsageGasConfig, usage: u64): u64 {
+        calculate_gas(config.target_usage, usage, &config.read_curve)
+    }
+
+    fun calculate_create_gas(config: &UsageGasConfig, usage: u64): u64 {
+        calculate_gas(config.target_usage, usage, &config.create_curve)
+    }
+
+    fun calculate_write_gas(config: &UsageGasConfig, usage: u64): u64 {
+        calculate_gas(config.target_usage, usage, &config.write_curve)
+    }
+
+    public(friend) fun on_reconfig() acquires StorageGas, StorageGasConfig {
+        assert!(
+            exists<StorageGasConfig>(@aptos_framework),
+            error::not_found(ESTORAGE_GAS_CONFIG)
+        );
+        assert!(
+            exists<StorageGas>(@aptos_framework),
+            error::not_found(ESTORAGE_GAS)
+        );
+        let (items, bytes) = state_storage::current_items_and_bytes();
+        let gas_config = borrow_global<StorageGasConfig>(@aptos_framework);
+        let gas = borrow_global_mut<StorageGas>(@aptos_framework);
+        gas.per_item_read = calculate_read_gas(&gas_config.item_config, items);
+        gas.per_item_create = calculate_create_gas(&gas_config.item_config, items);
+        gas.per_item_write = calculate_write_gas(&gas_config.item_config, items);
+        gas.per_byte_read = calculate_read_gas(&gas_config.byte_config, bytes);
+        gas.per_byte_create = calculate_create_gas(&gas_config.byte_config, bytes);
+        gas.per_byte_write = calculate_write_gas(&gas_config.byte_config, bytes);
+    }
+
+    #[test(framework = @aptos_framework)]
+    fun test_initialize_and_reconfig(framework: signer) acquires StorageGas, StorageGasConfig {
+        state_storage::initialize(&framework);
+        initialize(&framework);
+        on_reconfig();
+        let gas_parameter = borrow_global<StorageGas>(@aptos_framework);
+        assert!(gas_parameter.per_item_read == 100, 0);
+        assert!(gas_parameter.per_item_create == 100, 0);
+        assert!(gas_parameter.per_item_write == 100, 0);
+        assert!(gas_parameter.per_byte_read == 1, 0);
+        assert!(gas_parameter.per_byte_create == 1, 0);
+        assert!(gas_parameter.per_byte_write == 1, 0);
+    }
+
+    fun test_curve() {
+        let constant_curve = new_gas_curve(5, 5, vector[]);
+        let linear_curve = new_gas_curve(1, 1000, vector[]);
+        let standard_curve = base_32_exponential_curve(1, 1000);
+        let i = 7;
+        let old_standard_curve_gas = 1;
+        while (i <= BASIS_POINT_DENOMINATION + 10) {
+            assert!(calculate_gas(BASIS_POINT_DENOMINATION, i, &constant_curve) == 5, 0);
```

Comment:
I would test the cases of both `target < BASIS_POINT_DENOMINATION` and `target > BASIS_POINT_DENOMINATION`